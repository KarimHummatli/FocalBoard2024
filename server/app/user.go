package app

import "github.com/mattermost/focalboard/server/model"

func (a *App) GetTeamUsers(teamID string, asGuestID string) ([]*model.User, error) {
	return a.store.GetUsersByTeam(teamID, asGuestID)
}

func (a *App) SearchTeamUsers(teamID string, searchQuery string, asGuestID string) ([]*model.User, error) {
	return a.store.SearchUsersByTeam(teamID, searchQuery, asGuestID)
}

func (a *App) UpdateUserConfig(userID string, patch model.UserPropPatch) (map[string]interface{}, error) {
	if err := a.store.PatchUserProps(userID, patch); err != nil {
		return nil, err
	}

	user, err := a.store.GetUserByID(userID)
	if err != nil {
		return nil, err
	}

	return user.Props, nil
}

func (a *App) UserIsGuest(userID string) (bool, error) {
	user, err := a.store.GetUserByID(userID)
	if err != nil {
		return false, err
	}
	return user.IsGuest, nil
}

func (a *App) CanSeeUser(seerUser string, seenUser string) (bool, error) {
	isGuest, err := a.UserIsGuest(seerUser)
	if err != nil {
		return false, err
	}
	if isGuest {
		hasSharedChannels, err := a.store.CanSeeUser(seerUser, seenUser)
		if err != nil {
			return false, err
		}
		return hasSharedChannels, nil
	}
	return true, nil
}
