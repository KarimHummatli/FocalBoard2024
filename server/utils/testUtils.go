package utils

import "github.com/stretchr/testify/mock"

var Anything = mock.MatchedBy(func(any) bool { return true })
