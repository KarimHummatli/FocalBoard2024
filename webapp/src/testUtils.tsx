// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IntlProvider} from 'react-intl'
import React from 'react'

export const wrapIntl = (children: any) => <IntlProvider locale='en'>{children}</IntlProvider>
