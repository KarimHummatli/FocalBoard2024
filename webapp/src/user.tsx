// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'

interface IUser {
    id: string,
    username: string,
    email: string,
    props: Record<string, any>,
    createAt: number,
    updateAt: number,
}

type WorkspaceUsers = {
    users: Array<IUser>
    usersById: Map<string, IUser>
}

export {IUser, WorkspaceUsers}
