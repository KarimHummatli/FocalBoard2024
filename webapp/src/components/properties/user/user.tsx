// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useContext} from 'react'
import Select from 'react-select'

import {IUser, WorkspaceUsersContext, WorkspaceUsersContextData} from '../../../user'

import './user.scss'
import {getSelectBaseStyle} from '../../../theme'

type Props = {
    value: string,
    readonly: boolean,
    onChange: (value: string) => void,
    tooltip?: string
}

const UserProperty = (props: Props): JSX.Element => {
    const workspaceUsers = useContext<WorkspaceUsersContextData>(WorkspaceUsersContext)
    const tooltipClassName = props.tooltip ? 'octo-tooltip tooltip-top' : ''

    if (props.readonly) {
        const userClassName = `UserProperty octo-propertyvalue ${tooltipClassName}`
        return (
            <div
                className={userClassName}
                data-tooltip={props.tooltip}
            >
                {workspaceUsers.usersById.get(props.value)?.username || props.value}</div>)
    }

    return (
        <Select
            options={workspaceUsers.users}
            isSearchable={true}
            isClearable={true}
            backspaceRemovesValue={true}
            className={'UserProperty'}
            styles={getSelectBaseStyle()}
            getOptionLabel={(o: IUser) => o.username}
            getOptionValue={(a: IUser) => a.id}
            value={workspaceUsers.usersById.get(props.value) || null}
            onChange={(item, action) => {
                if (action.action === 'select-option') {
                    props.onChange(item?.id || '')
                } else if (action.action === 'clear') {
                    props.onChange('')
                }
            }}
        />
    )
}

export default UserProperty
