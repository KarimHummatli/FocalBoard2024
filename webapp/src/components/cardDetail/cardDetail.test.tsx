// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react'

import 'isomorphic-fetch'
import {act, render} from '@testing-library/react'

import {FetchMock} from '../../test/fetchMock'
import {TestBlockFactory} from '../../test/testBlockFactory'
import {MutableBoardTree} from '../../viewModel/boardTree'
import {wrapIntl} from '../../testUtils'

import {MutableCardTree} from '../../viewModel/cardTree'

import {IUser} from '../../user'

import CardDetail from './cardDetail'

global.fetch = FetchMock.fn

beforeEach(() => {
    FetchMock.fn.mockReset()
})

describe('components/cardDetail/CardDetail', () => {
    const board = TestBlockFactory.createBoard()

    const view = TestBlockFactory.createBoardView(board)
    view.sortOptions = []
    view.groupById = undefined
    view.hiddenOptionIds = []

    const card = TestBlockFactory.createCard(board)

    const createdAt = Date.parse('01 Jan 2021 00:00:00 GMT')
    const comment1 = TestBlockFactory.createCard(board)
    comment1.type = 'comment'
    comment1.title = 'Comment 1'
    comment1.parentId = card.id
    comment1.createAt = createdAt

    const comment2 = TestBlockFactory.createCard(board)
    comment2.type = 'comment'
    comment2.title = 'Comment 2'
    comment2.parentId = card.id
    comment2.createAt = createdAt

    const cardTree = new MutableCardTree(card)
    cardTree.comments = [comment1, comment2]

    test('should show comments', async () => {
        FetchMock.fn.mockReturnValueOnce(FetchMock.jsonResponse(JSON.stringify([board, view, card, comment1, comment2])))
        FetchMock.fn.mockReturnValue(FetchMock.jsonResponse(JSON.stringify({username: 'username_1'} as IUser)))
        const boardTree = await MutableBoardTree.sync(board.id, view.id, {})
        expect(boardTree).not.toBeUndefined()

        const component = wrapIntl(
            <CardDetail
                boardTree={boardTree!}
                cardTree={cardTree}
                readonly={false}
            />,
        )

        let container

        await act(async () => {
            const result = render(component)
            container = result.container
        })

        expect(container).toMatchSnapshot()
    })

    test('should show comments in readonly view', async () => {
        FetchMock.fn.mockReturnValueOnce(FetchMock.jsonResponse(JSON.stringify([board, view, card, comment1, comment2])))
        FetchMock.fn.mockReturnValue(FetchMock.jsonResponse(JSON.stringify({username: 'username_1'} as IUser)))
        const boardTree = await MutableBoardTree.sync(board.id, view.id, {})
        expect(boardTree).not.toBeUndefined()

        const component = wrapIntl(
            <CardDetail
                boardTree={boardTree!}
                cardTree={cardTree}
                readonly={true}
            />,
        )

        let container

        await act(async () => {
            const result = render(component)
            container = result.container
        })

        expect(container).toMatchSnapshot()
    })
})
