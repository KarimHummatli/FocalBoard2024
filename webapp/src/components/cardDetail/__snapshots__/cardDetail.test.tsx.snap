// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/cardDetail/CardDetail should show comments 1`] = `
<div>
  <div
    class="CardDetail content"
  >
    <div
      class="BlockIconSelector"
    >
      <div
        class="MenuWrapper"
      >
        <div
          class="octo-icon size-l"
        >
          <span>
            i
          </span>
        </div>
      </div>
    </div>
    <div
      class="EditableAreaWrap"
    >
      <textarea
        class="EditableArea Editable title"
        height="0"
        placeholder="Untitled"
        rows="1"
        spellcheck="true"
        title="title"
      >
        title
      </textarea>
      <div
        class="EditableAreaContainer"
      >
        <textarea
          aria-hidden="true"
          class="EditableAreaReference Editable title"
          dir="auto"
          disabled=""
          rows="1"
        >
          title
        </textarea>
      </div>
    </div>
    <div
      class="octo-propertylist CardDetailProperties"
    >
      <div
        class="octo-propertyrow"
      >
        <div
          class="MenuWrapper"
        >
          <div
            class="octo-propertyname"
          >
            <button
              class="Button    "
              type="button"
            >
              <span>
                Property 1
              </span>
            </button>
          </div>
        </div>
        <div
          class="octo-propertyvalue"
          tabindex="0"
        >
          <span
            class="Label propColorBrown "
          >
            value 1
          </span>
        </div>
      </div>
      <div
        class="octo-propertyrow"
      >
        <div
          class="MenuWrapper"
        >
          <div
            class="octo-propertyname"
          >
            <button
              class="Button    "
              type="button"
            >
              <span>
                Property 2
              </span>
            </button>
          </div>
        </div>
        <div
          class="octo-propertyvalue"
          tabindex="0"
        >
          <span
            class="Label empty "
          >
            Empty
          </span>
        </div>
      </div>
      <div
        class="octo-propertyrow"
      >
        <div
          class="MenuWrapper"
        >
          <div
            class="octo-propertyname"
          >
            <button
              class="Button    "
              type="button"
            >
              <span>
                Property 3
              </span>
            </button>
          </div>
        </div>
        <div
          class="octo-propertyvalue"
          tabindex="0"
        >
          <span
            class="Label empty "
          >
            Empty
          </span>
        </div>
      </div>
      <div
        class="octo-propertyname add-property"
      >
        <button
          class="Button    "
          type="button"
        >
          <span>
            + Add a property
          </span>
        </button>
      </div>
    </div>
    <hr />
    <div
      class="CommentsList"
    >
      <div
        class="Comment comment"
      >
        <div
          class="comment-header"
        >
          <img
            class="comment-avatar"
            src="data:image/svg+xml,<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\" style=\\"fill: rgb(192, 192, 192);\\"><rect width=\\"100\\" height=\\"100\\" /></svg>"
          />
          <div
            class="comment-username"
          >
            username_1
          </div>
          <div
            class="comment-date"
          >
            January 01, 5:30 AM
          </div>
          <div
            class="MenuWrapper"
          >
            <div
              class="Button IconButton"
            >
              <i
                class="CompassIcon icon-dots-horizontal OptionsIcon"
              />
            </div>
          </div>
        </div>
        <div
          class="comment-text"
        >
          <p>
            Comment 1
          </p>
        </div>
      </div>
      <div
        class="Comment comment"
      >
        <div
          class="comment-header"
        >
          <img
            class="comment-avatar"
            src="data:image/svg+xml,<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\" style=\\"fill: rgb(192, 192, 192);\\"><rect width=\\"100\\" height=\\"100\\" /></svg>"
          />
          <div
            class="comment-username"
          />
          <div
            class="comment-date"
          >
            January 01, 5:30 AM
          </div>
          <div
            class="MenuWrapper"
          >
            <div
              class="Button IconButton"
            >
              <i
                class="CompassIcon icon-dots-horizontal OptionsIcon"
              />
            </div>
          </div>
        </div>
        <div
          class="comment-text"
        >
          <p>
            Comment 2
          </p>
        </div>
      </div>
      <div
        class="commentrow"
      >
        <img
          class="comment-avatar"
          src="data:image/svg+xml,<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\" style=\\"fill: rgb(192, 192, 192);\\"><rect width=\\"100\\" height=\\"100\\" /></svg>"
        />
        <div
          class="MarkdownEditor octo-editor newcomment "
        >
          <div
            class="octo-editor-preview octo-placeholder"
          >
            <p>
              Add a comment...
            </p>
          </div>
          <div
            class="octo-editor-activeEditor"
            style="visibility: hidden; position: absolute; top: 0px; left: 0px;"
          >
            <div
              id="simplemde-editor-1-wrapper"
            >
              <textarea
                id="simplemde-editor-1"
                style="display: none;"
              />
              <div
                class="EasyMDEContainer"
              >
                <div
                  class="CodeMirror cm-s-easymde CodeMirror-wrap"
                >
                  <div
                    style="overflow: hidden; position: relative; width: 3px; height: 0px;"
                  >
                    <textarea
                      autocapitalize="off"
                      autocorrect="off"
                      spellcheck="false"
                      style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"
                      tabindex="0"
                    />
                  </div>
                  <div
                    class="CodeMirror-vscrollbar"
                    cm-not-content="true"
                    tabindex="-1"
                  >
                    <div
                      style="min-width: 1px;"
                    />
                  </div>
                  <div
                    class="CodeMirror-hscrollbar"
                    cm-not-content="true"
                    tabindex="-1"
                  >
                    <div
                      style="height: 100%; min-height: 1px;"
                    />
                  </div>
                  <div
                    class="CodeMirror-scrollbar-filler"
                    cm-not-content="true"
                  />
                  <div
                    class="CodeMirror-gutter-filler"
                    cm-not-content="true"
                  />
                  <div
                    class="CodeMirror-scroll"
                    style="min-height: 10px;"
                    tabindex="-1"
                  >
                    <div
                      class="CodeMirror-sizer"
                      style="margin-left: 0px;"
                    >
                      <div
                        style="position: relative;"
                      >
                        <div
                          class="CodeMirror-lines"
                          role="presentation"
                        >
                          <div
                            role="presentation"
                            style="position: relative; outline: none;"
                          >
                            <div
                              class="CodeMirror-measure"
                            >
                              <pre
                                class="CodeMirror-line-like"
                              >
                                <span>
                                  xxxxxxxxxx
                                </span>
                              </pre>
                            </div>
                            <div
                              class="CodeMirror-measure"
                            />
                            <div
                              style="position: relative; z-index: 1;"
                            />
                            <div
                              class="CodeMirror-cursors"
                            />
                            <div
                              class="CodeMirror-code"
                              role="presentation"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      style="position: absolute; height: 50px; width: 1px;"
                    />
                    <div
                      class="CodeMirror-gutters"
                      style="display: none;"
                    />
                  </div>
                </div>
                <div
                  class="editor-preview-side editor-preview"
                />
              </div>
            </div>
          </div>
        </div>
        
      </div>
      <hr />
    </div>
  </div>
  <div
    class="CardDetail content fullwidth"
  >
    <div
      class="octo-content"
    >
      <div
        class="octo-block"
      >
        <div
          class="octo-block-margin"
        />
        <div
          class="MarkdownEditor octo-editor  "
        >
          <div
            class="octo-editor-preview octo-placeholder"
          >
            <p>
              Add a description...
            </p>
          </div>
          <div
            class="octo-editor-activeEditor"
            style="visibility: hidden; position: absolute; top: 0px; left: 0px;"
          >
            <div
              id="simplemde-editor-2-wrapper"
            >
              <textarea
                id="simplemde-editor-2"
                style="display: none;"
              />
              <div
                class="EasyMDEContainer"
              >
                <div
                  class="CodeMirror cm-s-easymde CodeMirror-wrap"
                >
                  <div
                    style="overflow: hidden; position: relative; width: 3px; height: 0px;"
                  >
                    <textarea
                      autocapitalize="off"
                      autocorrect="off"
                      spellcheck="false"
                      style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"
                      tabindex="0"
                    />
                  </div>
                  <div
                    class="CodeMirror-vscrollbar"
                    cm-not-content="true"
                    tabindex="-1"
                  >
                    <div
                      style="min-width: 1px;"
                    />
                  </div>
                  <div
                    class="CodeMirror-hscrollbar"
                    cm-not-content="true"
                    tabindex="-1"
                  >
                    <div
                      style="height: 100%; min-height: 1px;"
                    />
                  </div>
                  <div
                    class="CodeMirror-scrollbar-filler"
                    cm-not-content="true"
                  />
                  <div
                    class="CodeMirror-gutter-filler"
                    cm-not-content="true"
                  />
                  <div
                    class="CodeMirror-scroll"
                    style="min-height: 10px;"
                    tabindex="-1"
                  >
                    <div
                      class="CodeMirror-sizer"
                      style="margin-left: 0px;"
                    >
                      <div
                        style="position: relative;"
                      >
                        <div
                          class="CodeMirror-lines"
                          role="presentation"
                        >
                          <div
                            role="presentation"
                            style="position: relative; outline: none;"
                          >
                            <div
                              class="CodeMirror-measure"
                            >
                              <pre
                                class="CodeMirror-line-like"
                              >
                                <span>
                                  xxxxxxxxxx
                                </span>
                              </pre>
                            </div>
                            <div
                              class="CodeMirror-measure"
                            />
                            <div
                              style="position: relative; z-index: 1;"
                            />
                            <div
                              class="CodeMirror-cursors"
                            />
                            <div
                              class="CodeMirror-code"
                              role="presentation"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      style="position: absolute; height: 50px; width: 1px;"
                    />
                    <div
                      class="CodeMirror-gutters"
                      style="display: none;"
                    />
                  </div>
                </div>
                <div
                  class="editor-preview-side editor-preview"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="CardDetail content add-content"
  >
    <div
      class="MenuWrapper"
    >
      <button
        class="Button    "
        type="button"
      >
        <span>
          Add content
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`components/cardDetail/CardDetail should show comments in readonly view 1`] = `
<div>
  <div
    class="CardDetail content"
  >
    <div
      class="BlockIconSelector"
    >
      <div
        class="octo-icon size-l readonly"
      >
        <span>
          i
        </span>
      </div>
    </div>
    <div
      class="EditableAreaWrap"
    >
      <textarea
        class="EditableArea Editable readonly title"
        height="0"
        placeholder="Untitled"
        readonly=""
        rows="1"
        spellcheck="true"
        title="title"
      >
        title
      </textarea>
      <div
        class="EditableAreaContainer"
      >
        <textarea
          aria-hidden="true"
          class="EditableAreaReference Editable readonly title"
          dir="auto"
          disabled=""
          rows="1"
        >
          title
        </textarea>
      </div>
    </div>
    <div
      class="octo-propertylist CardDetailProperties"
    >
      <div
        class="octo-propertyrow"
      >
        <div
          class="octo-propertyname"
        >
          Property 1
        </div>
        <div
          class="octo-propertyvalue"
          tabindex="0"
        >
          <span
            class="Label propColorBrown "
          >
            value 1
          </span>
        </div>
      </div>
      <div
        class="octo-propertyrow"
      >
        <div
          class="octo-propertyname"
        >
          Property 2
        </div>
        <div
          class="octo-propertyvalue"
          tabindex="0"
        >
          <span
            class="Label empty "
          >
            Empty
          </span>
        </div>
      </div>
      <div
        class="octo-propertyrow"
      >
        <div
          class="octo-propertyname"
        >
          Property 3
        </div>
        <div
          class="octo-propertyvalue"
          tabindex="0"
        >
          <span
            class="Label empty "
          >
            Empty
          </span>
        </div>
      </div>
    </div>
    <hr />
    <div
      class="CommentsList"
    >
      <div
        class="Comment comment"
      >
        <div
          class="comment-header"
        >
          <img
            class="comment-avatar"
            src="data:image/svg+xml,<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\" style=\\"fill: rgb(192, 192, 192);\\"><rect width=\\"100\\" height=\\"100\\" /></svg>"
          />
          <div
            class="comment-username"
          />
          <div
            class="comment-date"
          >
            January 01, 5:30 AM
          </div>
        </div>
        <div
          class="comment-text"
        >
          <p>
            Comment 1
          </p>
        </div>
      </div>
      <div
        class="Comment comment"
      >
        <div
          class="comment-header"
        >
          <img
            class="comment-avatar"
            src="data:image/svg+xml,<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\" style=\\"fill: rgb(192, 192, 192);\\"><rect width=\\"100\\" height=\\"100\\" /></svg>"
          />
          <div
            class="comment-username"
          />
          <div
            class="comment-date"
          >
            January 01, 5:30 AM
          </div>
        </div>
        <div
          class="comment-text"
        >
          <p>
            Comment 2
          </p>
        </div>
      </div>
      <hr />
    </div>
  </div>
  <div
    class="CardDetail content fullwidth"
  >
    <div
      class="octo-content"
    >
      <div
        class="octo-block"
      >
        <div
          class="octo-block-margin"
        />
      </div>
    </div>
  </div>
</div>
`;
